include/sdsl/rank_support_int.hpp

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2016, the SDSL Project Authors.  All rights reserved.
+// Please see the AUTHORS file for details.  Use of this source code is governed
+// by a BSD license that can be found in the LICENSE file.
+/*! \file rank_support_int.hpp
+    \brief rank_support_int.hpp contains classes that support a sdsl::int_vector with constant time rank information.
+           Rank is defined as the number of occurrences of a value up to a given position.
+	\author Christopher Pockrandt
+*/
+#ifndef INCLUDED_SDSL_RANK_SUPPORT_INT
+#define INCLUDED_SDSL_RANK_SUPPORT_INT
+
+/** \defgroup rank_support_group Rank Support (RS)
+ * This group contains data structures which support an sdsl::int_vector with the rank method.
+ */
+
+#include "int_vector.hpp"
+#include "uint128_t.hpp"
+
+// TODO: benchmark the use of compiler hints for branch prediction
+#define likely(x)       __builtin_expect((x),1)
+#define unlikely(x)     __builtin_expect((x),0)
+
+//! Namespace for the succinct data structure library.
+namespace sdsl {
+
+//! The base class of classes supporting rank_queries for a sdsl::int_vector in constant time.
+/*!
+*/
+
+constexpr size_t floor_log2(size_t const n)
+{
+    return (n == 1) ? 0 : 1 + floor_log2(n >> 1);
+}
+
+constexpr size_t ceil_log2(size_t const n)
+{
+    return (n == 1) ? 0 : floor_log2(n - 1) + 1;
+}
+
+template <uint8_t alphabet_size>
+class rank_support_int {
+
+public:
+	typedef typename int_vector<>::size_type size_type;
+	typedef typename int_vector<>::value_type value_type;
+
+    static_assert(alphabet_size > 2, "Rank support is only implemented on int_vectors with an alphabet size of > 2.");
+
+protected:
+
+    // Constructs a bit mask with the pattern w of a given length.
+    // It is concatenated until the length of the bitmask reaches max_length.
+    template <typename uintX_t>
+    static constexpr uintX_t bm_rec(const uintX_t w, const uint8_t length, const uint8_t max_length)
+    {
+        return (length >= max_length) ? w : bm_rec(w | (w << length), length << 1, max_length);
+    }
+
+protected:
+	const int_vector<>* m_v; //!< Pointer to the rank supported bit_vector
+    static constexpr uint8_t t_v{alphabet_size};
+    static constexpr uint8_t t_b{ceil_log2(alphabet_size)};
+	static constexpr uint64_t even_mask{bm_rec<uint64_t>(bits::lo_set[t_b], t_b * 2, 64)};
```

Comment:
So this function produces (for dna4): 0011'0011'0011...
Maybe you can add a small comment.